src/main/java/io/jenkins/plugins/credentials/secretsmanager/AwsCredentials.java

```diff
@@ -0,0 +1,173 @@
+abstract class AwsCredentials extends BaseStandardCredentials implements StringCredentials, StandardUsernamePasswordCredentials, SSHUserPrivateKey, StandardCertificateCredentials {
+
+    private static final char[] EMPTY_PASSWORD = {};
+    private static final Secret NONE = Secret.fromString("");
+
+    static final String USERNAME_TAG = "jenkins:credentials:username";
+
+    private final Map<String, String> tags;
+
+    AwsCredentials(String id, String description, Map<String, String> tags) {
+        super(id, description);
+        this.tags = tags;
+    }
+
+    public Map<String, String> getTags() {
+        return tags;
+    }
+
+    @Nonnull
+    @Override
+    public Secret getSecret() {
+        return Secret.fromString(getSecretString());
+    }
+
+    @NonNull
+    @Override
+    public Secret getPassword() {
+        if (tags.containsKey(USERNAME_TAG)) {
+            // username/password
+            return Secret.fromString(getSecretString());
+        } else {
+            // certificate
+            return NONE;
+        }
+    }
+
+    @NonNull
+    @Override
+    public String getUsername() {
+        if (tags.containsKey(USERNAME_TAG)) {
+            return tags.get(USERNAME_TAG);
+        } else {
+            throw new CredentialsUnavailableException("username", Messages.noUsernameError());
+        }
+    }
+
+    @Override
+    public Secret getPassphrase() {
+        return NONE;
+    }
+
+    @NonNull
+    @Override
+    public List<String> getPrivateKeys() {
+        return Collections.singletonList(getPrivateKey());
+    }
+
+    @NonNull
+    @Deprecated
+    @Override
+    public String getPrivateKey() {
+        final String secretValue = getSecretString();
+
+        if (SshKeyValidator.isValid(secretValue)) {
+            return secretValue;
+        } else {
+            throw new CredentialsUnavailableException("privateKey", Messages.noPrivateKeyError());
+        }
+    }
+
+    @NonNull
+    @Override
+    public KeyStore getKeyStore() {
+        final byte[] secretValue = getSecretBinary();
+
+        try (InputStream stream = new ByteArrayInputStream(secretValue)) {
+            final KeyStore keyStore = KeyStore.getInstance(KeyStore.getDefaultType());
+            // JDK9 workaround: PKCS#12 keystores must have at least an empty password (not null)
+            keyStore.load(stream, EMPTY_PASSWORD);
+            return keyStore;
+        } catch (IOException | CertificateException | KeyStoreException | NoSuchAlgorithmException ex) {
+            throw new CredentialsUnavailableException("keyStore", Messages.noCertificateError());
+        }
+    }
+
+    private String getSecretString() {
+        return getSecretValue().match(new SecretValue.Matcher<String>() {
+            @Override
+            public String string(String str) {
+                return str;
+            }
+
+            @Override
+            public String binary(byte[] bytes) {
+                return null;
```

Comment:
I'm not sure of the API spec, but is it possible to return something here? If it's not expected to be called, would it make sense to throw an exception?

Reply:
Thinking in terms of K-maps, an AWS Secret that contains neither a string nor a binary value should be a “can’t happen” state: AWS would have slipped up badly on their input validation for that to occur. I guess it would be prudent to throw an IllegalStateException then

Reply:
annotated getSecretValue() with `@NonNull`, threw exception if the AWS call ever gets to this state